diff --git a/src/HttpDownloadCommand.cc b/src/HttpDownloadCommand.cc
index 0011a443..4acc64a7 100644
--- a/src/HttpDownloadCommand.cc
+++ b/src/HttpDownloadCommand.cc
@@ -53,6 +53,8 @@
 #include "SinkStreamFilter.h"
 #include "util.h"
 #include "SocketRecvBuffer.h"
+#include "LogFactory.h"
+#include "message.h"
 
 namespace aria2 {
 
@@ -122,6 +124,15 @@ bool HttpDownloadCommand::prepareForNextSegment()
     if (lastOffset == range.endByte + 1) {
       return prepareForRetry(0);
     }
+    else if (range.getContentLength() == getFileEntry()->getLength()) {
+      // Now we check if the body is empty. If it is, we need to retry
+      // the request.
+      if (httpResponse_->getContentLength() == 0) {
+        A2_LOG_WARN(fmt(MSG_STRANGE_RANGE_HEADER, getCuid(), range.startByte,
+                        range.endByte, range.entityLength));
+        return prepareForRetry(0);
+      }
+    }
   }
 
   return DownloadCommand::prepareForNextSegment();
diff --git a/src/message.h b/src/message.h
index 4705d7fb..d0155f4b 100644
--- a/src/message.h
+++ b/src/message.h
@@ -99,6 +99,7 @@
 #define MSG_INCOMING_PEER_CONNECTION                            \
   "CUID#%" PRId64 " - Incoming connection, adding new command CUID#%" PRId64 ""
 #define MSG_ACCEPT_FAILURE "CUID#%" PRId64 " - Error in accepting connection"
+#define MSG_STRANGE_RANGE_HEADER "CUID#%" PRId64 " Strange range header. Response: %" PRId64 "-%" PRId64 "/%" PRId64 ""
 #define MSG_TRACKER_RESPONSE_PROCESSING_FAILED                  \
   "CUID#%" PRId64 " - Error occurred while processing tracker response."
 #define MSG_DHT_ENABLED_PEER "CUID#%" PRId64 " - The peer is DHT-enabled."
